src/Dialect/Mlir/VectorMachineSupport.cpp

```diff
@@ -0,0 +1,254 @@
+
+#include "src/Dialect/Mlir/VectorMachineSupport.hpp"
+#include "mlir/IR/BuiltinTypes.h"
+#include <algorithm>
+
+#define DEBUG_TYPE "dialect_builder"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+// =============================================================================
+// Handling of global vector machine support pointer
+
+/*static*/ VectorMachineSupport
+    *VectorMachineSupport::globalVectorMachineSupport = nullptr;
+
+/*static*/ void VectorMachineSupport::setGlobalVectorMachineSupport(
+    std::string arch, std::string cpu, std::string attr) {
+  // IBM Z servers use mcpu.
+  if (cpu.compare("z14") == 0) {
+    globalVectorMachineSupport = new Z14VectorMachineSupport();
+  } else if (cpu.compare("z15") == 0) {
+    globalVectorMachineSupport = new Z15VectorMachineSupport();
+  } else if (cpu.compare("z16") == 0) {
+    globalVectorMachineSupport = new Z16VectorMachineSupport();
+  } else if (arch.compare("x86-64") == 0) {
+    // Intel arch
+    if (cpu.compare("skylake") == 0 && attr.compare("avx2") == 0)
+      globalVectorMachineSupport = new AVX2x86VectorMachineSupport();
+    else
+      // Default seems to be SSE
+      globalVectorMachineSupport = new SSE42x86VectorMachineSupport();
+  } else {
+    // Unknown: disable
+    globalVectorMachineSupport = new NoVectorMachineSupport();
+  }
+  assert(globalVectorMachineSupport &&
+         "failed to allocate vector machine support");
+}
+
+/*static*/ void VectorMachineSupport::clearGlobalVectorMachineSupport() {
+  if (!globalVectorMachineSupport)
+    return;
+  delete globalVectorMachineSupport;
+  globalVectorMachineSupport = nullptr;
+}
+
+// =============================================================================
+// Methods shared among all VectorMachineSupport classes and subclasses
+
+int64_t VectorMachineSupport::getVectorLength(Type elementType) {
+  int64_t simdBitSize = getVectorBitWidth();
+  int64_t typeBitSize = elementType.getIntOrFloatBitWidth();
+  assert(simdBitSize >= typeBitSize && simdBitSize % typeBitSize == 0 &&
+         "bad machine vector length");
+  return (simdBitSize / typeBitSize);
+}
+
+double VectorMachineSupport::getAvgVectorLength(ArrayRef<GenericOps> &gops,
+    ArrayRef<int64_t> &gopsNum, Type elementType, int64_t &vectorizedOpNum,
+    int64_t &scalarOpNum) {
+  int64_t gopsSize = gops.size();
+  int64_t gopsNumSize = gopsNum.size();
+  int64_t totVL = 0.0;
+  vectorizedOpNum = 0;
+  scalarOpNum = 0;
+  // Determine which operations support SIMD and accumulate their vector
+  // lengths.
+  for (int64_t i = 0; i < gopsSize; ++i) {
+    // Even if unsupported, we can still proceed with one value per op.
+    int64_t vl = std::max((int64_t)1, getVectorLength(gops[i], elementType));
+    // If past last value, assume 1; otherwise use actual value.
+    int64_t num = 1;
+    if (i < gopsNumSize)
+      num = gopsNum[i];
+    // Accumulate weighted scalar/vectorized num and vl length.
+    if (vl > 0)
```

Comment:
yes, vl == 0 means no simd, so increase the count for scalar op. I now renamed `totVl` to `totProcessedValues`. Processed values are vl when vl>0, and 1 otherwise (for scalar). That should be clearer.

Thanks for catching that.